spring-data-neo4j-rx/src/main/java/org/neo4j/springframework/data/core/Neo4jTemplate.java

```diff
@@ -311,55 +311,61 @@ public void deleteAll(Class<?> domainType) {
-				.getPersistentEntity(associationTargetType);
-
-			Collection<RelationshipDescription> relationships = neo4jPersistentEntity.getRelationships();
-			RelationshipDescription relationship = relationships.stream()
-				.filter(r -> r.getFieldName().equals(inverse.getName()))
-				.findFirst().get();
+			Neo4jPersistentEntity<?> targetNodeDescription = neo4jMappingContext
+				.getPersistentEntity(ctx.getAssociationTargetType());
 
+			Object fromId = propertyAccessor.getProperty(neo4jPersistentEntity.getRequiredIdProperty());
 			// remove all relationships before creating all new if the entity is not new
 			// this avoids the usage of cache but might have significant impact on overall performance
 			if (!neo4jPersistentEntity.isNew(parentObject)) {
 				Statement relationshipRemoveQuery = cypherGenerator.createRelationshipRemoveQuery(neo4jPersistentEntity,
-					relationship, targetNodeDescription.getPrimaryLabel());
+					ctx.getRelationship(), targetNodeDescription.getPrimaryLabel());
 
 				neo4jClient.query(renderer.render(relationshipRemoveQuery))
 					.bind(fromId).to(FROM_ID_PARAMETER_NAME).run();
 			}
 
-			if (value == null) {
+			if (ctx.inverseValueDoesNotExists()) {
 				return;
 			}
 
-			for (Object relatedValue : Relationships.unifyRelationshipValue(inverse, value)) {
-
-				Object valueToBeSaved = relatedValue instanceof Map.Entry ?
-					((Map.Entry) relatedValue).getValue() :
-					relatedValue;
-				valueToBeSaved = eventSupport.maybeCallBeforeBind(valueToBeSaved);
+			for (Object relatedValue : Relationships.unifyRelationshipValue(ctx.getInverse(), ctx.getValue())) {
 
-				Long relatedInternalId = saveRelatedNode(valueToBeSaved, associationTargetType, targetNodeDescription);
+				// here map entry is not always anymore a dynamic association
+				Object valueToBeSaved = relatedValue;
+				if (relatedValue instanceof Map.Entry && ctx.getInverse().isDynamicAssociation()) {
+					valueToBeSaved = ((Map.Entry) relatedValue).getValue();
+				} else if (relatedValue instanceof Map.Entry && ctx.getRelationship().hasRelationshipProperties()) {
+					valueToBeSaved = ((Map.Entry) relatedValue).getKey();
+				}
 
-				Statement relationshipCreationQuery = cypherGenerator
-					.createRelationshipCreationQuery(neo4jPersistentEntity,
-						relationship,
-						relatedValue instanceof Map.Entry ? ((Map.Entry<String, ?>) relatedValue).getKey() : null,
-						relatedInternalId);
+				valueToBeSaved = eventSupport.maybeCallBeforeBind(valueToBeSaved);
 
-				neo4jClient.query(renderer.render(relationshipCreationQuery))
-					.bind(fromId).to(FROM_ID_PARAMETER_NAME).run();
+				Long relatedInternalId = saveRelatedNode(valueToBeSaved, ctx.getAssociationTargetType(),
+					targetNodeDescription);
+
+				// handle creation of relationship depending on properties on relationship or not
+				RelationshipStatementHolder queryHolder = ctx.getRelationship().hasRelationshipProperties()
```

Comment:
If it is an instance of a StatementHolder it should also be named `statementHolder`.